harness/determined/launch/autodeepspeed.py

```diff
@@ -0,0 +1,236 @@
+            cert=cert,
+        )
+
+        # Wrap it in a pid_server to ensure that we can't hang if a worker fails.
+        # This is useful for deepspeed which does not have good error handling for remote processes
+        # spun up by pdsh.
+        pid_server_cmd = [
+            "python3",
+            "-m",
+            "determined.exec.pid_server",
+            "--on-fail",
+            "SIGTERM",
+            "--on-exit",
+            "SIGTERM",
+            "--grace-period",
+            "3",
+            f"/tmp/pid_server-{info.allocation_id}",
+            str(len(info.slot_ids)),
+            "--",
+        ]
+
+        logging.debug(
+            f"Non-chief [{info.container_rank}] training process launch "
+            f"command: {run_sshd_command}."
+        )
+        return subprocess.Popen(pid_server_cmd + run_sshd_command).wait()
+
+    # We always need to set this variable to initialize the context correctly, even in the single
+    # slot case.
+    os.environ["USE_DEEPSPEED"] = "1"
+
+    # The chief has several layers of wrapper processes:
+    # - a top-level pid_server, which causes the whole container to exit if any local worker dies.
+    # - deepspeed, which launches $slots_per_trial copies of the following layers:
+    #     - a pid_client process to contact the local pid_server
+    #     - worker_process_wrapper, which redirects stdin/stdout to the local container
+    #     - harness.py, which actually does the training for the worker
+
+    pid_server_cmd = [
+        "python3",
+        "-m",
+        "determined.exec.pid_server",
+        "--on-fail",
+        "SIGTERM",
+        "--on-exit",
+        "SIGTERM",
+        "--grace-period",
+        "5",
+        f"/tmp/pid_server-{info.allocation_id}",
+        str(len(info.slot_ids)),
+        "--",
+    ]
+
+    hostfile_path = "/tmp/hostfile.txt"
+    master_address = create_hostlist_file(
+        hostfile_path=pathlib.Path(hostfile_path),
+        num_proc_per_machine=len(info.slot_ids),
+        ip_addresses=info.container_addrs,
+    )
+    cmd = create_run_command(master_address, hostfile_path)
+
+    pid_client_cmd = [
+        "python3",
+        "-m",
+        "determined.exec.pid_client",
+        f"/tmp/pid_server-{info.allocation_id}",
+        "--",
+    ]
+
+    log_redirect_cmd = [
+        "python3",
+        "-m",
+        "determined.exec.worker_process_wrapper",
+        "RANK",
+        "--",
+    ]
+
+    harness_cmd = [
+        "python3",
+        "-m",
+        "determined.exec.harness",
+        "--train-entrypoint",
+        train_entrypoint,
+    ]
+
+    logging.debug(f"chief worker calling deepspeed with args: {cmd[1:]} ...")
+
+    full_cmd = pid_server_cmd + cmd + pid_client_cmd + log_redirect_cmd + harness_cmd
+
+    multi_machine = len(info.container_addrs) > 1
+    if not multi_machine:
+        return subprocess.Popen(full_cmd).wait()
+
+    # Chief worker also needs to run sshd when using pdsh and multi-machine training.
```

Comment:
ok, I know this is my second nit on the same line, but maybe move it above the `sshd_process =` line?